python/cuml/ensemble/randomforest_shared.pyx

```diff
@@ -0,0 +1,132 @@
+
+cdef extern from "Python.h":
+    Py_buffer* PyMemoryView_GET_BUFFER(PyObject* mview)
+
+cdef class PyBufferFrameWrapper:
+    cdef PyBufferFrame _handle
+    cdef Py_ssize_t shape[1]
+    cdef Py_ssize_t strides[1]
+
+    def __cinit__(self):
+        pass
+
+    def __dealloc__(self):
+        pass
+
+    def __getbuffer__(self, Py_buffer* buffer, int flags):
+        cdef Py_ssize_t itemsize = self._handle.itemsize
+
+        self.shape[0] = self._handle.nitem
+        self.strides[0] = itemsize
+
+        buffer.buf = self._handle.buf
+        buffer.format = self._handle.format
+        buffer.internal = NULL
+        buffer.itemsize = itemsize
+        buffer.len = self._handle.nitem * itemsize
+        buffer.ndim = 1
+        buffer.obj = self
+        buffer.readonly = 0
+        buffer.shape = self.shape
+        buffer.strides = self.strides
+        buffer.suboffsets = NULL
+
+    def __releasebuffer__(self, Py_buffer *buffer):
+        pass
+
+cdef PyBufferFrameWrapper MakePyBufferFrameWrapper(PyBufferFrame handle):
+    cdef PyBufferFrameWrapper wrapper = PyBufferFrameWrapper()
+    wrapper._handle = handle
+    return wrapper
+
+cdef list _get_frames(ModelHandle model):
+    frames = []
+    cdef Model* model_obj = <Model*>model
+    cdef vector[PyBufferFrame] interface = model_obj.GetPyBuffer()
+    cdef vector[PyBufferFrame].iterator it = interface.begin()
+    while it != interface.end():
+        v = deref(it)
+        w = MakePyBufferFrameWrapper(v)
+        frames.append(memoryview(w))
+        inc(it)
+    return frames
+
+cdef ModelHandle _init_from_frames(vector[PyBufferFrame] frames) except *:
+    cdef Model* model_obj = new Model()
+    model_obj.InitFromPyBuffer(frames)
+    cdef ModelHandle handle = <ModelHandle>model_obj
+    return handle
+
+
+def get_frames(model: uintptr_t) -> List[memoryview]:
+    return _get_frames(<ModelHandle> model)
+
+
+def init_from_frames(frames: List[Union[bytes, memoryview]],
+                     format_str: List[str], itemsize: List[int]) -> uintptr_t:
+    cdef vector[PyBufferFrame] cpp_frames
+    cdef Py_buffer* buf
+    cdef PyBufferFrame cpp_frame
+    for i, frame in enumerate(frames):
+        if isinstance(frame, memoryview):
+            buf = PyMemoryView_GET_BUFFER(<PyObject*>frame)
+            cpp_frame.buf = buf.buf
+        else:
+            cpp_frame.buf = <char*>(frame)
+        cpp_frame.format = format_str[i]
+        cpp_frame.itemsize = itemsize[i]
+        cpp_frame.nitem = len(frame) // itemsize[i]
+        cpp_frames.push_back(cpp_frame)
+    return <uintptr_t> _init_from_frames(cpp_frames)
+
+
+def treelite_serialize(
+    model: uintptr_t
+) -> Dict[str, Union[List[str], List[bytes]]]:
+    frames = get_frames(model)
+    header = {'format_str': [x.format.encode('utf-8') for x in frames],
+              'itemsize': [x.itemsize for x in frames]}
+    return {'header': header, 'frames': [bytes(x) for x in frames]}
```

Comment:
Would be better to keep these as `memoryview`s. Constructing a `bytes` object copies. Dask can handle the `memoryview`s directly anyways.

```suggestion
    return {'header': header, 'frames': frames}
```

Reply:
Pickle doesn't handle memoryview yet.

Reply:
We'll want to refactor the RF object to avoid pickling in the future. However, I'd like to keep this PR a limited scope. I'll remove the use of pickling in a follow-up PR.

Reply:
How is pickling involved?

An alternative that is pickle-friendly and doesn't require a copy would be a NumPy array. Could do something like this for example (would need to `import numpy as np` in this file too):

```suggestion
    return {'header': header, 'frames': [np.asarray(x) for x in frames]}
```

Reply:
The RF object has pickling method defined.

Reply:
Applying your suggestion about using NumPy arrays to avoid copying. This should keep the pickle module happy.

Reply:
No worries. Just didn't realize we were already pickling.

FWIW there is another benefit of using NumPy arrays here. Namely they support out-of-band serialization (analogous to Dask's serialization) via pickle protocol 5 out-of-the-box. So this means that one can do zero-copy serialization that way as well :)